Subcomponents handling (#230)
Veridise/circom: llzk_backend/docs/subcomponents.md

# Subcomponent lowering

This document gives a high-level overview of how subcomponents are lowered to
LLZK.

## 1. Subcomponent detection

During the *declaration info* phase the backend collects all the subcomponent
declarations and the instances assigned to them. For each declaration the
backend collects the name of the subcomponent and the dimensions if it was
declared as an array of subcomponents.

For each instance the backend collects the type that was instantiated for that
subcomponent. Currently the backend only supports scalar subcomponents or array

> subcomponents in which all instances are of the same type.

Reviewer: ```suggestion
During the *declaration info* phase, the backend collects all the subcomponent
declarations and the instances assigned to them. For each declaration, the
backend collects the name of the subcomponent and the dimensions if it was
declared as an array of subcomponents.

For each instance, the backend collects the type that was instantiated for that
subcomponent. Currently the backend only supports scalar subcomponents or array
subcomponents in which all instances are of the same type.
```

```
// Supported 
component a;
a = A(X);

component b[N];
b[0] = B(X);
b[1] = B(X);

// Not yet supported
component c[N];
c[0] = C(X);
c[1] = C(Y);
```

For each subcomponent detected in this phase the backend creates a field in the
struct and adds the name to the scope in preparation for lowering.

## 2. Subcomponent lowering

### Subcomponent constructors

Call expressions that create a new subcomponent are translated to calls to
`@compute`/`@constrain`. Since the call is an expression, it needs to return a
value. In the case of `@compute` that is the operation's result and in the case

> of `@constrain` is the first argument of the function. Refering to a

Reviewer: ```suggestion
of `@constrain` that is the first argument of the function. Refering to a
```

subcomponent by name is refering to this value. The second piece of the
construction is an assign var statement. When a variable tied to a subcomponent
gets assigned it does different things depending on the function. In `@compute`

> updates the its value with the result of lowering the right hand expression. In

Reviewer: ```suggestion
gets assigned it does different things depending on the function. In `@compute`,
it updates the value with the result of lowering the right hand expression. In
```
